net/nat/nat.c

```diff
@@ -0,0 +1,341 @@
+/****************************************************************************
+ * net/nat/nat.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <debug.h>
+
+#include "icmp/icmp.h"
+#include "nat/nat.h"
+#include "tcp/tcp.h"
+#include "udp/udp.h"
+
+#ifdef CONFIG_NET_NAT
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nat_port_select_without_stack
+ *
+ * Description:
+ *   Select an available port number for TCP/UDP protocol, or id for ICMP.
+ *   Used when corresponding stack is disabled.
+ *
+ * Input Parameters:
+ *   domain     - The domain of the packet.
+ *   protocol   - The L4 protocol of the packet.
+ *   ip         - The IP bind with the port (in network byte order).
+ *   portno     - The local port (in network byte order), as reference.
+ *
+ * Returned Value:
+ *   port number on success; 0 on failure
+ *
+ ****************************************************************************/
+
+#if (defined(CONFIG_NET_TCP) && defined(CONFIG_NET_TCP_NO_STACK))  || \
+    (defined(CONFIG_NET_UDP) && defined(CONFIG_NET_UDP_NO_STACK))  || \
+    (defined(CONFIG_NET_ICMP) && !defined(CONFIG_NET_ICMP_SOCKET))
+
+static uint16_t nat_port_select_without_stack(
+    uint8_t domain, uint8_t protocol, FAR const union ip_addr_u *ip,
+    uint16_t portno)
+{
+  uint16_t hport = NTOHS(portno);
+  while (nat_port_inuse(domain, protocol, ip, portno))
+    {
+      ++hport;
+      if (hport >= CONFIG_NET_DEFAULT_MAX_PORT ||
+          hport < CONFIG_NET_DEFAULT_MIN_PORT)
+        {
+          hport = CONFIG_NET_DEFAULT_MIN_PORT;
+        }
+
+      portno = HTONS(hport);
+    }
```

Comment:
Thanks for pointing out this, then TCP/UDP may also face the same problem (https://github.com/apache/nuttx/pull/12110)

@wangchen61698 @Meissi-jian If we set `CONFIG_NET_DEFAULT_MIN_PORT` near to `CONFIG_NET_DEFAULT_MAX_PORT`, e.g. setting to only allow 8 ports, then the busy loop may possibly happen after we have used all 8 ports, for both TCP/UDP/NAT. Maybe need to consider a way to avoid the busy loop, such as disallow the port to be set as `MIN_PORT` twice in one loop?

Reply:
Suggestion: before entering in the while loop, test if hport is inside the valid range, when inside the loop increase the hport and compare it with CONFIG_NET_DEFAULT_MAX_PORT, if it is greater or equal, then break the loop and report and error.

Note: I think there is not a POSIX errno to indicate port already in use, only addresses: EADDRINUSE.

Reply:
merge port range logic to common function?

```
      ++hport;
      if (hport >= CONFIG_NET_DEFAULT_MAX_PORT ||
          hport < CONFIG_NET_DEFAULT_MIN_PORT)
        {
          hport = CONFIG_NET_DEFAULT_MIN_PORT;
        }

      portno = HTONS(hport);
```

Reply:
I tried to add a commit to optimize this, please review the newly added commit, thanks.